connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java

```diff
@@ -724,9 +696,19 @@ private static Object convertToConnect(Schema schema, JsonNode jsonValue) {
                     break;
             }
         }
-        typeConverter = TO_CONNECT_CONVERTERS.get(schemaType);
+
+        final JsonToConnectTypeConverter typeConverter = TO_CONNECT_CONVERTERS.get(schemaType);
         if (typeConverter == null)
-            throw new DataException("Unknown schema type: " + schema.type());
+            throw new DataException("Unknown schema type: " + schemaType);
+
+        if (jsonValue.isNull()) {
+            assert schema != null; // otherwise we would have thrown above because of unknown schema type
+            if (schema.isOptional())
+                return null;
+            if (schema.defaultValue() != null)
+                return schema.defaultValue(); // any logical type conversions should already have been applied
```

Comment:
The order of considering optional vs default value is different than before. Was that intentional?